### PR TITLE
 sqlgen: a random generator producer

### PR DESCRIPTION
**Pull request by tangenta**
```
.
├── goyacc
│   └── main.go
├── main.go
├── README.md
└── sqlgen
    ├── bnf.go                     # contains production datastructure and lexer for goyacc.
    ├── bnf_parser.go              # generated from goyacc.
    ├── bnf_parser_test.go
    ├── bnf.y                      # bnf grammar.
    ├── generator.go               # random generator genrator.
    ├── generator_lib.go           # a library for generator.go, implement the random behavior.
    ├── generator_types.go         # types for generator.
    ├── plugin_maxloop_limit.go    # a plugin for detecting and limiting max recursive production.
    ├── sample                     # a sample generated from this tool.
    │   ├── declarations.go
    │   ├── sample_test.go
    │   └── start.go               # bnf grammar in golang representation.
    ├── sample_bnf.txt             # sample bnf grammar.
    ├── util.go                    # BuildProductionMap, ParseYaccFile and utilties.
    └── util_test.go

3 directories, 17 files

```

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @bb7133 @winkyao @lonng

File: sqlgen/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# sqlgen
+
+sqlgen is a tool for generating a random generator. 
+
+After the execution, a golang package named *pgkname* is generated, which contains a set of files. The `Generate()` locating in *production*.go can be used to generate a random string, according to the syntax described in *bnf*.
```
Comment (kennytm) [inline sqlgen/README.md]:
```suggestion
After the execution, a golang package named *pkgname* is generated, which contains a set of files. The `Generate()` locating in *production*.go can be used to generate a random string, according to the syntax described in *bnf*.
```

**Review by kennytm — COMMENTED** [review]
What is a "random generator"? Do you mean "random SQL statement generator"?

**Comment by tangenta** [comment]
@kennytm It can be used to generate a "random SQL statement generator".
There is a bug in this PR. After I fix it, please take a look.

**Comment by tangenta** [comment]
PTAL @bb7133 @winkyao @lonng @kennytm 
I am not sure if I should give a talk about this tool.

File: sqlgen/bnf_helper/main.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"github.com/pingcap/tidb-tools/sqlgen/sqlgen"
+	"log"
+	"os"
```
Comment (zimulala) [inline sqlgen/bnf_helper/main.go:20]:
```suggestion
	"bufio"
	"flag"
	"log"
	"os"
	
	"github.com/pingcap/tidb-tools/sqlgen/sqlgen"
```

**Review by DQinYuan — APPROVED** [review]
LGTM

**Comment by kennytm** [comment]
We have got https://github.com/pingcap/go-randgen already.

/close

**Pull request closed without merging**